src/trackinfoobject.cpp

```diff
@@ -738,23 +739,51 @@ QString TrackInfoObject::getURL() {
 }
 
 ConstWaveformPointer TrackInfoObject::getWaveform() {
-    return m_waveform;
+    return m_pWaveform;
 }
 
 void TrackInfoObject::setWaveform(ConstWaveformPointer pWaveform) {
-    m_waveform = pWaveform;
+    if (m_pWaveform == pWaveform) {
```

Comment:
The comparison of the pointers is thread-safe, but the whole set function is 
not atomic. If we acquire the lock after the comparison in very rare cases 
the member pointer might be modified by another thread in the meantime.

On 01/12/2016 09:50 AM, Daniel Schürmann wrote:
>
> In src/trackinfoobject.cpp 
> <https://github.com/mixxxdj/mixxx/pull/841#discussion_r49427732>:
>
> >  }
> >
> >  void TrackInfoObject::setWaveform(ConstWaveformPointer pWaveform) {
> > -    m_waveform = pWaveform;
> > +    if (m_pWaveform == pWaveform) {
>
> The check has to be done inside the lock.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/841/files#r49427732>.
>